Resources/doc/configuration.rst
Configuration
-------------

Publishers configuration
========================

TaskPublisher can have multiple publishers configured.
When event is triggered it will loop trough each configured publisher and execute publishers ``publish`` method.

**Publishers**
==============

AMQP publisher
==============

To enable AMQP publisher with default values:

.. code-block:: yaml

    ongr_task_messenger:
        publishers:
            default:
                amqp: ~


With custom values:

.. code-block:: yaml

    ongr_task_messenger:
        publishers:
            default:
                amqp:
                    host: 192.168.1.1
                    ...



========== ============================================= ===========================
**Option** **Default value**                              **Description**
---------- --------------------------------------------- ---------------------------
factory    ongr_task_messenger.simple_connection_factory Factory service
publisher  ongr_task_messenger.publisher.amqp            Publisher service
class      PhpAmqpLib\Connection\AMQPConnection          Class used to create connection object
host       127.0.0.1                                     AMQP host
port       5672                                          AMQP port
user       guest                                         Login user
password   guest                                         Login password
========== ============================================= ===========================


Beanstalkd publisher
====================

To enable beanstalkd publisher with default values:

.. code-block:: yaml

    ongr_task_messenger:
        publishers:
            default:
                beanstalkd: ~


With custom values:

.. code-block:: yaml

    ongr_task_messenger:
        publishers:
            default:
                beanstalkd:
                    host: 192.168.1.1
                    ...


========== ============================================= ===========================
**Option** **Default value**                              **Description**
---------- --------------------------------------------- ---------------------------
factory    ongr_task_messenger.simple_connection_factory Factory service
publisher  ongr_task_messenger.publisher.beanstalkd      Publisher service
class      Pheanstalk\Pheanstalk                         Class used to create connection object
host       127.0.0.1                                     beanstalkd host
port       11300                                         beanstalkd port
user       null                                          Login user
password   null                                          Login password
========== ============================================= ===========================


Custom publisher
================

One can configure custom publisher and pass additional parameters to constructor.
Publishers name node must be ``custom``.

.. code-block:: yaml

    ongr_task_messenger:
        publishers:
            foo_publisher:
                custom:
                    publisher: acme_test.publisher.redis
                    factory: acme_test.redis_connection_factory
                    class: 'Predis\Client'
                    host: 127.0.0.1
                    port: 6379
                    user: guest
                    password: guest
                    arguments:
                        scheme: tcp


========== ============================================= ======== ======================================
**Option** **Default value**                             Required **Description**
---------- --------------------------------------------- -------- --------------------------------------
factory                                                  yes      Factory service
publisher                                                yes      Publisher service
class                                                    yes      Class used to create connection object
host       127.0.0.1                                     no       host address
port                                                     yes      host port
user       null                                          no       Login user
password   null                                          no       Login password
========== ============================================= ======== ======================================


Log level
=========

To specify publishers logging level set ``ongr_task_messenger.log_level`` parameter or in your ``app\config\config.yml``
Valid values are PSR3 log levels.

Example:

.. code-block:: yaml

    ongr_task_messenger:
        log_level: debug


Example bundle configuration
============================

For full bundle configuration example see `config_test.yml <https://github.com/ongr-io/TaskMessengerBundle/blob/master/Tests/app/config/config_test.yml>`_

Running tests
=============

Bundle should be installed with ``require-dev`` dependencies.

To run bundles tests ``RabbitMQ``, ``beanstalkd`` and ``redis`` services must be installed. You can find setup example in ``.travis.yml``.


DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the ONGR package.
 *
 * (c) NFQ Technologies UAB <info@nfq.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace ONGR\TaskMessengerBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\NodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;

/**
 * This is the class that validates and merges configuration from app/config files.
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('ongr_task_messenger');
        $rootNode
            ->children()
                ->scalarNode('log_level')
                    ->defaultValue('debug')
                ->end()
                ->append($this->getPublishersNode())
            ->end();

        return $treeBuilder;
    }

    /**
     * Publishers configuration node.
     *
     * @return NodeDefinition
     * @throws InvalidConfigurationException
     */
    private function getPublishersNode()
    {
        $builder = new TreeBuilder();
        $node = $builder->root('publishers');

        $node
            ->info('Defines task publishers manager node.')
                ->prototype('array')
                ->info('Defines publishers and configuration settings.')
                    ->children()
                        ->arrayNode('amqp')
                            ->children()
                                ->scalarNode('class')
                                    ->defaultValue('PhpAmqpLib\Connection\AMQPConnection')
                                ->end()
                                ->scalarNode('factory')
                                    ->defaultValue('ongr_task_messenger.simple_connection_factory')
                                ->end()
                                ->scalarNode('publisher')
                                    ->defaultValue('ongr_task_messenger.publisher.amqp')
                                ->end()
                                ->scalarNode('host')
                                    ->defaultValue('127.0.0.1')
                                ->end()
                                ->scalarNode('port')
                                    ->defaultValue(5672)
                                ->end()
                                ->scalarNode('user')
                                    ->defaultValue('guest')
                                ->end()
                                ->scalarNode('password')
                                    ->defaultValue('guest')
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('beanstalkd')
                            ->children()
                                ->scalarNode('factory')
                                    ->defaultValue('ongr_task_messenger.simple_connection_factory')
                                ->end()
                                ->scalarNode('publisher')
                                    ->defaultValue('ongr_task_messenger.publisher.beanstalkd')
                                ->end()
                                ->scalarNode('class')
                                    ->defaultValue('Pheanstalk\Pheanstalk')
                                ->end()
                                ->scalarNode('host')
                                    ->defaultValue('127.0.0.1')
                                ->end()
                                ->scalarNode('port')
                                    ->defaultValue(11300)
                                ->end()
                                ->scalarNode('user')
                                    ->defaultValue(null)
                                ->end()
                                ->scalarNode('password')
                                    ->defaultValue(null)
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('custom')
                            ->children()
                                ->scalarNode('factory')
                                    ->isRequired()
                                ->end()
                                ->scalarNode('class')
                                    ->isRequired()
                                ->end()
                                ->scalarNode('publisher')
                                    ->isRequired()
                                ->end()
                                ->scalarNode('host')
                                    ->defaultValue('127.0.0.1')
                                ->end()
                                ->scalarNode('port')
                                    ->isRequired()
                                ->end()
                                ->scalarNode('user')
                                    ->defaultValue(null)
                                ->end()
                                ->scalarNode('password')
                                    ->defaultValue(null)
                                ->end()
                                ->variableNode('arguments')
                                    ->defaultValue([])
                                ->end()
                            ->end()
                        ->end()
                    ->end()
            ->end();

        return $node;
    }
}
